searchlib/src/vespa/searchlib/tensor/hnsw_index_saver.cpp

```diff
@@ -15,32 +39,37 @@ HnswIndexSaver::HnswIndexSaver(const HnswGraph &graph)
     _meta_data.entry_docid = entry.docid;
     _meta_data.entry_level = entry.level;
     size_t num_nodes = graph.node_refs.size();
-    _meta_data.nodes.reserve(num_nodes);
+    assert (num_nodes <= 0xfffffffful);
+    size_t linkCount = countValidLinks(graph);
+    assert (linkCount <= 0xfffffffful);
+    _meta_data.refs.reserve(linkCount);
+    _meta_data.nodes.reserve(num_nodes+1);
     for (size_t i = 0; i < num_nodes; ++i) {
-        LevelVector node;
+        _meta_data.nodes.push_back(_meta_data.refs.size());
         auto node_ref = graph.node_refs[i].load_acquire();
         if (node_ref.valid()) {
             auto levels = graph.nodes.get(node_ref);
             for (const auto& links_ref : levels) {
-                auto level = links_ref.load_acquire();
-                node.push_back(level);
+                _meta_data.refs.push_back(links_ref.load_acquire());
             }
         }
-        _meta_data.nodes.emplace_back(std::move(node));
     }
+    _meta_data.nodes.push_back(_meta_data.refs.size());
 }
 
 void
 HnswIndexSaver::save(BufferWriter& writer) const
 {
     writer.write(&_meta_data.entry_docid, sizeof(uint32_t));
     writer.write(&_meta_data.entry_level, sizeof(int32_t));
-    uint32_t num_nodes = _meta_data.nodes.size();
+    uint32_t num_nodes = _meta_data.nodes.size() - 1;
     writer.write(&num_nodes, sizeof(uint32_t));
-    for (const auto &node : _meta_data.nodes) {
-        uint32_t num_levels = node.size();
+    for (uint32_t i(0); i < num_nodes; i++) {
+        uint32_t offset = _meta_data.nodes[i];
+        uint32_t num_levels = _meta_data.nodes[i+1] - offset;
```

Comment:
Consider caching `_meta_data.nodes[i + 1]` in a separate variable, e.g. `next_offset`

Reply:
Fixed